Support timestamps on WI console

mokjasmine opened the pull request:
#### d69d0f7bc8d83054032900f13091b03d9a724212
<pre>
Support timestamps on WI console
<a href="https://bugs.webkit.org/show_bug.cgi?id=248240">https://bugs.webkit.org/show_bug.cgi?id=248240</a>

Reviewed by Patrick Angle.

We want a toggle-able flag in web inspector&apos;s console for showing a timestamp per console message.
Send the timestamp from the backend and display per console message. For duplicate messages, keep the repeat count and update the timestamp to that of the latest console message.
Collapse/expand functionality for duplicate messages when toggling off/on timestamps will come in a future PR.

-LayoutTests/inspector/console/console-api-expected.txt
-LayoutTests/inspector/console/console-api.html
-LayoutTests/inspector/worker/console-basic-expected.txt
-LayoutTests/inspector/worker/console-basic.html
-LayoutTests/inspector/console/console-table-expected.txt
-LayoutTests/inspector/console/console-table.html
-Source/JavaScriptCore/inspector/ConsoleMessage.cpp
-Source/JavaScriptCore/inspector/protocol/Console.json
-Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js
-Source/WebInspectorUI/UserInterface/Base/Setting.js
-Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js
-Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js
-Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js
-Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css
-Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js
-Source/WebInspectorUI/UserInterface/Views/LogContentView.js
-Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js

Canonical link: <a href="https://commits.webkit.org/258675@main">https://commits.webkit.org/258675@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c59e867feb597e3189a8aaedbbc62617de1ea80a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111711 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171921 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12621 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2468 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94716 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109429 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92872 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37311 "Build was cancelled. Recent messages:Failed to print configuration; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Pull request contains relevant changes; Skipped layout-tests; 23 flakes 87 failures; 11 flakes 74 failures; Reverted pull request changes; Compiled WebKit (cancelled)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91536 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79063 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92696 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5040 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25775 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88989 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2718 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2216 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29577 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11216 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45268 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91914 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5960 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6933 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20580 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Base/Setting.js:
```diff
@@ -222,6 +222,7 @@ WI.settings = {
     showRulers: new WI.Setting("show-rulers", false),
     showRulersDuringElementSelection: new WI.Setting("show-rulers-during-element-selection", true),
     showScopeChainOnPause: new WI.Setting("show-scope-chain-sidebar", true),
+    showTimestamps: new WI.Setting("show-timestamps", false),
```
We should be more specific with this setting name in case timestamps become toggle able in some other part of Web Inspector some day. Maybe `showConsoleMessageTimestamps` and `show-console-message-timestamps`?

patrickangle commented on Source/WebCore/page/PageConsoleClient.cpp:
```diff
@@ -189,8 +189,9 @@ void PageConsoleClient::messageWithTypeAndLevel(MessageType type, MessageLevel l
+    auto timestamp = MonotonicTime::now().secondsSinceEpoch();
 
-    auto message = makeUnique<Inspector::ConsoleMessage>(MessageSource::ConsoleAPI, type, level, messageText, arguments.copyRef(), lexicalGlobalObject);
+    auto message = makeUnique<Inspector::ConsoleMessage>(MessageSource::ConsoleAPI, type, level, messageText, arguments.copyRef(), lexicalGlobalObject, 0, Seconds(timestamp));
```
I wonder if instead we should change Inspector::ConsoleMessage to automatically create a timestamp at its creation so that all the other locations that create console messages can take advantage of the timestamp being present (a search for `makeUnique<Inspector::ConsoleMessage>` turns up a handful of different places a message may be created).

dcrousso replied:
yeah if we're gonna do this we want _all_ messages to have timestamps, not just those that come from `console.*` APIs

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
```diff
@@ -310,6 +310,11 @@
 
+.console-message .timestamp {
+    color: hsl(0, 0%, 53%);
+    padding-right: 5px;
```
We prefer `padding-inline-end` so that the padding automatically ends up in the correct place for RTL languages.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css:
```diff
@@ -310,6 +310,11 @@
 }
 
+.console-message .timestamp {
+    color: hsl(0, 0%, 53%);
```
This should probably be `color: var(--text-color-tertiary);` so that it matches other, similar text and allows the color to be adjusted for light/dark mode.

dcrousso replied:
also please make sure this looks good in dark mode :)

hopefully `--text-color-tertiary` will do TheRightThing™️

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -39,6 +39,7 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
         this._message = message;
         this._expandable = false;
         this._repeatCount = message._repeatCount || 0;
+        this._timestamp = message._timestamp || 0;
```
Nit: I'd just use `this._message.timestamp` in this file instead.

Also, I know the above line breaks this rule, but we should generally not access a "private" member of a class from outside that class (we use the underscore prefix to denote a member that should be treated as private).

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +162,37 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+        milliseconds = (milliseconds < 100 && milliseconds >= 10) ? "0" + milliseconds : milliseconds;
+        milliseconds = (milliseconds < 10) ? "00" + milliseconds : milliseconds;
+
+        this._timestampElement.textContent = hours + ":" + minutes + ":" + seconds + ":" + milliseconds;
```
Converting seconds to a human-readable timestamp would probably make a for a good utility method (similar to our existing `secondsToString` utility function).

Also, this isn't resilient to various localizations. For example in the US we would generally show the hours as a number from 1-12, with an AM/PM after the time. I think what we might actually want to do here is something along the lines of
```
let timeFormat = new Intl.DateTimeFormat("default", {hour: "2-digit", minute: "2-digit", second: "2-digit", fractionalSecondDigits: 3})

this._timestampElement.textContent = timeFormat.format(this._message.timestamp)
```
Which will follow locale-specific rules for showing the hours+minutes+seconds+milliseconds. We could probably make the time format a static member, like `WI.ConsoleMessageView.TimestampFormat` and reuse it for each message.

dcrousso replied:
could also maybe just use `Date.prototype.toLocaleString` 

also, we'd probably want `hour12: false` :)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -99,6 +100,8 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
         this._appendStackTrace();
 
         this._renderRepeatCount();
+        if (WI.settings.showTimestamps.value)
```
Should `renderTimestamp` check this setting itself so it can remove the element if necessary?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +162,37 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+        }
+
+        if (!this._timestampElement) {
+            this._timestampElement = document.createElement("span");
```
The constructor should set `this._timestampElement = null` to prevent us from "reshaping" this class at runtime. It also helps us keep track of what the various member variables are in a class if the constructor explicitly sets them (even if it sets them to null).

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -89,6 +89,14 @@ WI.LogContentView = class LogContentView extends WI.ContentView
 
+        // 'Show Timestamp' toggle
+        this._showTimestampsNavigationItem = new WI.CheckboxNavigationItem("show-timestamps", WI.UIString("Show Timestamps", "Show Timestamps @ Console", "Checkbox shown in the Console to show timestamps."), WI.settings.showTimestamps.value);
+        this._showTimestampsNavigationItem.tooltip = WI.UIString("Show timestamp for each console log");
```
It seems like we generally use "console log" to refer to the messages as a whole, so maybe "console message" would be more accurate here?

dcrousso replied:
does this really even need a tooltip?  it's already showing pretty self-explanatory text

definitely agree with @patrickangle about the "console message" part tho if we do decide to keep a tooltip :)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +162,37 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+        }
+
+        let date = new Date(this._timestamp * 1000);
+        var hours = date.getHours(),
```
Nit: `let`

dcrousso replied:
also please do a `let` for each instead of using `,`

patrickangle reviewed (commented):
Nice work! A few general comments/observations:

It's odd in your screenshot that only some messages show the timestamp. Since the timestamp should always be sent with a message, I'd expect the toggle to show/hide all timestamps.

There appear to be a Web Inspector test failing on macOS (inspector/console/console-table.html) that probably just needs to be modified to filter out the timestamp, similar to how `_stackTrace` is already filtered for the message.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +162,37 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+            this._element.insertBefore(this._timestampElement, this._element.firstChild);
+        }
+
+        let date = new Date(this._timestamp * 1000);
```
the timestamp never changes, right?  if so, can we just early `return` if we've already created the `this._timestampElement`?  i.e. we shouldn't have to (re)generate a string for the timestamp if we've already done so

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -89,6 +89,14 @@ WI.LogContentView = class LogContentView extends WI.ContentView
             this._emulateUserGestureNavigationItemGroup = new WI.GroupNavigationItem([this._emulateUserGestureNavigationItem, new WI.DividerNavigationItem]);
         }
 
+        // 'Show Timestamp' toggle
```
unnecessary comment.  please remove

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -89,6 +89,14 @@ WI.LogContentView = class LogContentView extends WI.ContentView
         }
 
+        // 'Show Timestamp' toggle
+        this._showTimestampsNavigationItem = new WI.CheckboxNavigationItem("show-timestamps", WI.UIString("Show Timestamps", "Show Timestamps @ Console", "Checkbox shown in the Console to show timestamps."), WI.settings.showTimestamps.value);
```
part of me feels like this should be in the Console pane of the Settings Tab since it's probably not super likely that a developer will want to toggle this on/off frequently

@patrickangle wdyt?

patrickangle replied:
Unless we (@dcrousso, @mokjasmine) feel there is a compelling use case for being able to quick toggle these on/off, that would be fine with me.

dcrousso replied:
yeah thinking on this more I'd probably lean towards moving this to the Console pane of the Settings Tab.

maybe we'd just have a "Show:" with a "[ ] Timestamps" item next to it (kinda like what we do for whitespace)?

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Console.json:
```diff
@@ -127,7 +127,8 @@
             "parameters": [
-                { "name": "count", "type": "integer", "description": "New repeat count value." }
+                { "name": "count", "type": "integer", "description": "New repeat count value." },
+                { "name": "timestamp", "type": "number" }
```
The `timestamp` parameter must be marked as optional to maintain compatibility with older backends (for example a Mac inspecting an older iOS device).

e.g. `{ "name": "timestamp", "type": "number", "optional": true }`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
 
+    set timestamp(timestamp)
+    {
+        this._message._timestamp = timestamp;
```
This breaks our general contract of now accessing private members of other classes. I know I previously commented that this view should just always get the timestamp from the message, but now that we are handling repeating messages, maybe this view should keep track of the timestamp like it does the repeat count.

Or if you keep the updating of the underlying WI.ConsoleMessage, you should add a `set timeout(newTimeout)` to `ConsoleMessage.js`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+
+    _renderTimestamp(renderlatestTimestamp = false)
+    {
+        if (!this._message._timestamp) {
```
This should use the public getter for timestamp (e.g. `this._message.timestamp`)

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+        }
+    
+        if (renderlatestTimestamp) {
+            let date = new Date(this._message._timestamp * 1000);
```
Ditto :176

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+            this._renderTimestamp(true);
+    }
+
+    _renderTimestamp(renderlatestTimestamp = false)
```
Style: names should be camel case, e.g. `renderLatestesTimestamp`. Also a falsey value is the default if the parameter is not provided, so no need for `= false`.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -39,10 +39,13 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
         // Any extra parameters will be displayed as children of this message.
         this._extraParameters = message.parameters;
+
+        this._timestampElement = null
```
Style: Missing trailing `;`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -99,6 +102,7 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
         this._appendStackTrace();
 
         this._renderRepeatCount();
+        this._renderTimestamp(true);            
```
Style: When passing a constant literal value into a function, we prefer to declare a const member to name the variable for clarity.

e.g.
```
const renderLatestTimestamp = true;
this._renderTimestamp(renderLatestTimestamp);
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+    {
+        this._message._timestamp = timestamp;
+        if (this._element)
+            this._renderTimestamp(true);
```
Ditto :105

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+            if (this._timestampElement) {
+                this._timestampElement.remove();
+                this._timestampElement = null;
+            }
```
We can simplify the nesting of this by using optional chaining without the `if (this._timestampElement)` e.g.
```
this._timestampElement?.remove();
this._timestampElement = null;
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+        }
+        if (!this._showTimestamps) {
+            if (this._timestampElement)
+                this._timestampElement.style.setProperty("display", "none");
```
This can also be simplified with optional chaining
```
this._timestampElement?.style.setProperty("display", "none");
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js:
```diff
@@ -159,6 +164,43 @@ WI.ConsoleMessageView = class ConsoleMessageView extends WI.Object
+            }
+            return;
+        }
+        if (!this._showTimestamps) {
```
A common pattern we use for lists of things that have dynamically showable/hideable content is for the content to always be present in the row (in this case the `ConsoleMessageView`, and let the view of all the rows handle observing the setting and applying a class to the list of console messages that hides/shows all of the timestamps via CSS. This way only one object has to have an observer for the setting changing, instead of every single console message having an observers, of which there might be thousands or more.

If you did this, it would also simplify this method as well as make it easier to understand where style comes from. You would no longer need to check the setting at all in `ConsoleMessageView`, instead making that the responsibility of its parent. It also allows us to implement a cleaner true CSS approach to showing/hiding the timestamps since we can just toggle a class like `shows-timestamps` on the parent view and let CSS handle what gets shown/hidden.

(Note there may still be times where we don't have a timestamp and can't create the element since Web Inspector supports connecting to older backends that may not have the information, so checking if a timestamp exists for the message is still necessary on a row-by-row basis like you did above on :176)

patrickangle reviewed (changes requested):
I have some concerns about the approach to observing the new setting in `ConsoleMessageView` which I noted below.

Also it looks like the `inspector/console/console-table.html` tests needs rebaselined for this change. We should make sure the timestamp is filtered since the time will vary for each test run.

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -330,6 +337,11 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            this._currentSessionOrGroup.element.removeAttribute("hide-timestamps");
+        else
+            this._currentSessionOrGroup.element.setAttribute("hide-timestamps", "true");
+
```
In new code we generally prefer using a CSS class for this, which also allows this to be reduced to a single line.
```
this._currentSessionOrGroup.element.classList.toggle("timestamps-visible", this._showTimestamps);
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -367,6 +379,21 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+    {
+        this._showTimestamps = WI.settings.showConsoleMessageTimestamps.value;
+        if (this._showTimestamps) {
+            this._currentSessionOrGroup.element.removeAttribute("hide-timestamps");
```
Ditto :340-344 (e.g. move this above the if statement and then you can get rid of the else below as well)

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -367,6 +379,21 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+                let consoleMessageView = this._consoleMessageViews[i];
+                if (consoleMessageView.constructor.name == "ConsoleMessageView")
+                    consoleMessageView.renderTimestamp();
+            }
```
When we don't need the index as part of the iteration it is preferable to use a for-of loop
```
for (let consoleMessageView of this._consoleMessageViews) {
    ...
}
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -367,6 +379,21 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+                    consoleMessageView.renderTimestamp();
+            }
+        }
+        else
```
Style: `else` belongs on the line of the last `if` clause's closing curly bracket.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSession.css:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2013 Apple Inc. All rights reserved.
```
`2023`?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSession.css:
```diff
@@ -0,0 +1,32 @@
+
+.console-session[hide-timestamps] .timestamp {
+    display: none;
+}
```
This style should probably be in `ConsoleMessageView.js` since that's the view where these styles are used. Also we prefer more specific selectors when possible. e.g. 
```
.console-message .timestamp {
    display: block;
}

.console-session:not(.timestamps-visible) .console-message .timestamp {
    display: none;
}
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
```diff
@@ -367,6 +379,21 @@ WI.JavaScriptLogViewController = class JavaScriptLogViewController extends WI.Ob
+            this._currentSessionOrGroup.element.removeAttribute("hide-timestamps");
+            for (let i = 0; i < this._consoleMessageViews.length; i++) {
+                let consoleMessageView = this._consoleMessageViews[i];
+                if (consoleMessageView.constructor.name == "ConsoleMessageView")
```
Can views other than `WI.ConsoleMessageView`s end up in this array? If not, there isn't a need to check this.

mokjasmine replied:
Yes other views have consistently ended up in this array

patrickangle replied:
👍 Sound good

Should have also mentioned yesterday that if keep the check we tend to prefer 
```
consoleMessageView instanceof WI.ConsoleMessageView
```
for type checks.

patrickangle commented on Source/JavaScriptCore/inspector/protocol/Console.json:
```diff
@@ -127,7 +127,8 @@
             "parameters": [
-                { "name": "count", "type": "integer", "description": "New repeat count value." }
+                { "name": "count", "type": "integer", "description": "New repeat count value." },
+                { "name": "timestamp", "type": "number", "optional": true }
```
Please include a description for the new parameter as well.

patrickangle reviewed (approved):
Awesome work - thank you for iterating on the patch!

Just the one comment below. Please also make sure all the bots pass the tests, it looks like [gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/36961) on the previous PR was failing, so we should make sure that that bot passes on this latest PR.

mokjasmine committed:
Support timestamps on WI console
https://bugs.webkit.org/show_bug.cgi?id=248240

Reviewed by Patrick Angle.

We want a toggle-able flag in web inspector's console for showing a timestamp per console message.
Send the timestamp from the backend and display per console message. For duplicate messages, keep the repeat count and update the timestamp to that of the latest console message.
Collapse/expand functionality for duplicate messages when toggling off/on timestamps will come in a future PR.

-LayoutTests/inspector/console/console-api-expected.txt
-LayoutTests/inspector/console/console-api.html
-LayoutTests/inspector/worker/console-basic-expected.txt
-LayoutTests/inspector/worker/console-basic.html
-LayoutTests/inspector/console/console-table-expected.txt
-LayoutTests/inspector/console/console-table.html
-Source/JavaScriptCore/inspector/ConsoleMessage.cpp
-Source/JavaScriptCore/inspector/protocol/Console.json
-Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js
-Source/WebInspectorUI/UserInterface/Base/Setting.js
-Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js
-Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js
-Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js
-Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.css
-Source/WebInspectorUI/UserInterface/Views/ConsoleMessageView.js
-Source/WebInspectorUI/UserInterface/Views/LogContentView.js
-Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js

Canonical link: https://commits.webkit.org/258675@main
